File: Python3/interview_questions/arrays/SmallestWindowToSort.py
import unittest

#   Given an array of integers that are out of order, determine the bound of the smallest
#   window that must be sorted in order for the entire array to be sorted.


def window(array):
    left = None
    right = None
    length = len(array)
    # float(inf) represents infinity or
    # negative infinity for a negative.
    max_seen = -float("inf")
    min_seen = float("inf")

    for i in range(length):
        # Get the max value seen so far.
        max_seen = max(max_seen, array[i])
        if array[i] < max_seen:
            right = i

    # Loop the array in reverse by 1.
    for i in range(length - 1, -1, -1):
        # Get the min value seen so far.
        min_seen = min(min_seen, array[i])
        if array[i] > min_seen:
            left = i

    return left, right


class TestWindow(unittest.TestCase):
    def test_window(self):
        array = window([3, 4, 7, 5, 6, 9])
        self.assertEqual(array[0], 2)
        self.assertEqual(array[1], 4)

    def test_window2(self):
        array = window([4, 1, 8, 9, 10])
        self.assertEqual(array[0], 0)
        self.assertEqual(array[1], 1)


if __name__ == "__main__":
    unittest.main()

File: Java/snippet/src/main/java/interview_questions/arrays/SmallestWindowToSort.java
package interview_questions.arrays;

public class SmallestWindowToSort {
  /*
  Given an array of integers that are out of order, determine the bound of the smallest
  window that must be sorted in order for the entire array to be sorted. For example,
  given [3, 5, 7, 6, 9], you should return (1, 3).
   */

  private static int left;
  private static int right;

  // Time complexity: O(N)
  // Space complexity: O(1)
  public static int[] window(int[] array) {
    // We set the minimum and maximum seen values
    // to to smallest and largest possible Integer values
    // in order to ensure Math.min and Math.max returns
    // the smallest and largest values when checking against
    // the current value in the array while looping.
    int max_seen = Integer.MIN_VALUE;
    int min_seen = Integer.MAX_VALUE;

    // Loop through the array.
    for (int i = 0; i < array.length; i++) {
      // Get the greatest current value since iteration started.
      max_seen = Math.max(max_seen, array[i]);

      // If the current value is less than then current maximum value seen
      // set the right bound to the index of the current value.
      if (array[i] < max_seen) {
        right = i;
      }
    }

    // Loop through the array in reverse.
    for (int i = array.length - 1; i >= 0; i--) {
      // Get the smallest current value since iteration started.
      min_seen = Math.min(min_seen, array[i]);

      // If the current value is greater than the current minimum value seen
      // set the right bound to the index of the current value.
      if (array[i] > min_seen) {
        left = i;
      }
    }

    // Return the left and right bounds.
    return new int[]{left, right};
  }

}
